cle/loader.py

```diff
@@ -787,6 +788,9 @@ def _possible_paths(self, spec):
             if self.main_object.binary is not None:
                 dirs.append(os.path.dirname(self.main_object.binary))
             if self._use_system_libs:
+                # Ideally this should be taken into account for each shared
+                # object, not just the main object.
+                dirs.extend(self.main_object.extra_load_path)
```

Comment:
I leave bigger refactoring to a different pull request. However the current design should be this easy.